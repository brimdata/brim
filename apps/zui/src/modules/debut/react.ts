import {selectGroup} from "./dom"

class ReactApi {
  constructor(public group: string) {}

  enter() {
    return selectGroup(this.group).enter()
  }

  exit() {
    return selectGroup(this.group).exit()
  }
}

export function useDebut(group: string) {
  return new ReactApi(group)
}
