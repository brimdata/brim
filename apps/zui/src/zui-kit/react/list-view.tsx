import {isEqual} from "lodash"
import React, {
  forwardRef,
  useEffect,
  useImperativeHandle,
  useMemo,
  useRef,
  useState,
} from "react"
import {FixedSizeList, ListChildComponentProps} from "react-window"
import {ListViewApi} from "../core"
import {createListView} from "../core/list-view/create-list-view"
import {defaultListViewState} from "../core/list-view/state"
import {ListViewArgs} from "../core/list-view/types"
import {ReactAdapterProps} from "./types"
import {useStateControllers} from "./use-state-controllers"
import {mergeRefs, useInitialScrollPosition, useOnScroll} from "./utils"
import classNames from "classnames"
import {useParentSize} from "src/app/core/hooks/use-parent-size"
import {TopShadow, useScrollShadow} from "src/views/load-pane/scroll-shadow"
import {call} from "src/util/call"

const padding = 8

export const InnerElement = forwardRef<any, any>(function Inner(props, ref) {
  const {style, ...rest} = props
  const height = `${parseFloat(style.height) + (padding ?? 0) * 2}px`

  return <div role="list" ref={ref} style={{...style, height}} {...rest} />
})

export const OuterElement = forwardRef<any, any>(function Outer(props, ref) {
  const {children, ...rest} = props
  const shadow = useScrollShadow(200)
  return (
    <div style={{height: "100%", width: "100%", position: "relative"}}>
      <TopShadow
        opacity={shadow.top}
        style={{
          height: "2px",
          background: "linear-gradient(rgba(0,0,0,0.25), transparent)",
        }}
      />
      <div
        {...rest}
        ref={mergeRefs(shadow.ref, ref)}
        onScroll={(e) => {
          call(rest.onScroll, e)
          shadow.onScroll()
        }}
      >
        {children}
      </div>
    </div>
  )
})

export const Row: React.ComponentType<
  React.PropsWithChildren<ListChildComponentProps>
> = React.memo(
  function Row({style, index, data}) {
    if (!data[index]) return null
    const {render, indent} = data[index]
    const innerStyle = {
      paddingLeft: 16 * (indent + 1),
      top: (style.top as number) + padding,
    }

    return (
      <div className="zed-view" style={{...style, ...innerStyle}}>
        {render}
      </div>
    )
  },
  (prev, next) => {
    return (
      prev.data[prev.index] === next.data[next.index] &&
      prev.index === next.index &&
      isEqual(prev.style, next.style)
    )
  }
)

export const ListView = forwardRef(function ListView(
  props: ListViewArgs & ReactAdapterProps,
  ref: React.Ref<ListViewApi>
) {
  const controllers = useStateControllers(props, defaultListViewState)
  const args = {...props, ...controllers}
  const outerRef = useRef<HTMLDivElement>()
  const list = useMemo(() => createListView(args), [props])
  const [rendered, setRendered] = useState({startIndex: 0, stopIndex: 10})
  list.rendered = rendered
  list.fill()
  useOnScroll(outerRef, list.onScroll.bind(list))
  useInitialScrollPosition(outerRef, props.initialScrollPosition)
  useImperativeHandle(ref, () => list, [list])
  useEffect(() => {
    list.element = outerRef.current
  }, [list])

  const {width, height} = useParentSize(outerRef)

  return (
    <FixedSizeList
      className={classNames(props.className, "zed-list-view")}
      innerRef={props.innerRef}
      height={props.height ?? height}
      width={props.width ?? width}
      outerRef={outerRef}
      itemCount={list.count}
      itemSize={20}
      itemData={[...list.rows]}
      itemKey={(i) => i.toString()}
      innerElementType={InnerElement}
      // outerElementType={OuterElement}
      overscanCount={8}
      onItemsRendered={(args) => {
        setRendered({
          startIndex: args.overscanStartIndex,
          stopIndex: args.overscanStopIndex,
        })
      }}
    >
      {Row}
    </FixedSizeList>
  )
})
