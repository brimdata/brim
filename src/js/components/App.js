/* @flow */

import {useDispatch, useSelector} from "react-redux"
import React, {useEffect} from "react"

import {ipcRenderer} from "electron"

import {XLatestError} from "./LatestError"
import AboutModal from "./AboutModal"
import ClusterGate from "./Login/ClusterGate"
import ErrorNotice from "./ErrorNotice"
import HTMLContextMenu from "./HTMLContextMenu"
import Preferences from "./Preferences/Preferences"
import Prefs from "../state/Prefs"
import View from "../state/View"
import brim from "../brim"
import refreshSpaceNames from "../flows/refreshSpaceNames"
import SpaceModal from "./SpaceModal"

export default function App() {
  brim.time.setZone(useSelector(View.getTimeZone))
  brim.time.setDefaultFormat(useSelector(Prefs.getTimeFormat))
  let dispatch = useDispatch()

  useEffect(() => {
    dispatch(refreshSpaceNames())
    ipcRenderer.invoke("windows:ready")
  }, [])

  return (
    <div className="app-wrapper">
      <div className="title-bar-drag-area" />
      <XLatestError />
      <ClusterGate />

      {/* Global Modals */}
      <ErrorNotice />
      <Preferences />
      <AboutModal />
      <SpaceModal />
      <HTMLContextMenu />
    </div>
  )
}
