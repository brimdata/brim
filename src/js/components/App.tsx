import {useSelector} from "react-redux"
import React, {useEffect} from "react"

import {ipcRenderer} from "electron"

import {XLatestError} from "./LatestError"
import AboutModal from "./AboutModal"
import ErrorNotice from "./ErrorNotice"
import HTMLContextMenu from "./HTMLContextMenu"
import Preferences from "./Preferences/Preferences"
import Prefs from "../state/Prefs"
import SpaceModal from "./SpaceModal"
import View from "../state/View"
import brim from "../brim"
import SearchPage from "./SearchPage"
import {PopNotice} from "./PopNotice"
import {Modals} from "./Modals"
import BrimToaster from "./BrimToaster"
import BrimTooltip from "./BrimTooltip"
import HookLog from "app/system-test/HookLog"
import ColumnsModal from "app/columns/columns-modal"
import AppRouter from "app/router/router"
import {Route, Switch} from "react-router"

export default function App() {
  brim.time.setZone(useSelector(View.getTimeZone))
  brim.time.setDefaultFormat(useSelector(Prefs.getTimeFormat))

  useEffect(() => {
    ipcRenderer.invoke("windows:ready")
    tabHistory.push(
      "/workspaces/localhost:9867/lakes/sp_1oOqildljMkbkd4qNg1y5nyWfSz/search"
    )
  }, [])

  return (
    <AppRouter>
      <Switch>
        <Route path="/">
          <div className="app-wrapper">
            <div className="title-bar-drag-area" />
            <XLatestError />
            <SearchPage />

            {/* Global Modals */}
            <ErrorNotice />
            <Preferences />
            <Modals />
            <AboutModal />
            <SpaceModal />
            <ColumnsModal />
            <HTMLContextMenu />
            <BrimToaster />
            <PopNotice />
            <BrimTooltip />
            {process.env.BRIM_ITEST === "true" && <HookLog />}
          </div>
        </Route>
      </Switch>
    </AppRouter>
  )
}
