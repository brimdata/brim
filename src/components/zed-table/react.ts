import {useState} from "react"
import {createZedTable} from "./create-zed-table"
import {ZedTableProps} from "./types"

export function useReactZedTable(opts: ZedTableProps) {
  const [api] = useState(() => ({
    current: createZedTable(opts),
  }))

  return api.current
}

// 1. Provide the state
// 2. Listen for when the state changes
// 3. Handle the appropriate state changes
// 4. Changes redux, re-renders the table
