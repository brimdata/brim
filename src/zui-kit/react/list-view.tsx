import {isEqual} from "lodash"
import React, {
  forwardRef,
  useImperativeHandle,
  useMemo,
  useRef,
  useState,
} from "react"
import {FixedSizeList, ListChildComponentProps} from "react-window"
import {ListViewApi} from "../core"
import {createListView} from "../core/list-view/create-list-view"
import {defaultListViewState} from "../core/list-view/state"
import {ListViewArgs} from "../core/list-view/types"
import {ReactAdapterProps} from "./types"
import {useStateControllers} from "./use-state-controllers"
import {useInitialScrollPosition, useOnScroll} from "./utils"

const padding = 8

export const InnerElement = forwardRef<any, any>(function InnerElement(
  {style, ...rest},
  ref
) {
  return (
    <div
      role="list"
      ref={ref}
      style={{
        ...style,
        height: `${parseFloat(style.height) + (padding ?? 0) * 2}px`,
      }}
      {...rest}
    />
  )
})

export const Row: React.ComponentType<
  React.PropsWithChildren<ListChildComponentProps>
> = React.memo(
  function Row({style, index, data}) {
    if (!data[index]) return null
    const {render, indent} = data[index]
    const innerStyle = {
      paddingLeft: 16 * (indent + 1),
      top: (style.top as number) + padding,
    }

    return (
      <div className="zed-view" style={{...style, ...innerStyle}}>
        {render}
      </div>
    )
  },
  (prev, next) => {
    return (
      prev.data === next.data &&
      prev.index === next.index &&
      isEqual(prev.style, next.style)
    )
  }
)

export const ListView = forwardRef(function ListView(
  props: ListViewArgs & ReactAdapterProps,
  ref: React.Ref<ListViewApi>
) {
  const controllers = useStateControllers(props, defaultListViewState)
  const args = {...props, ...controllers}
  const outerRef = useRef<HTMLDivElement>()
  const list = useMemo(() => createListView(args), [props])
  const [rendered, setRendered] = useState({startIndex: 0, stopIndex: 10})
  list.rendered = rendered
  list.fill()
  useOnScroll(outerRef, list.onScroll.bind(list))
  useInitialScrollPosition(outerRef, props.initialScrollPosition)
  useImperativeHandle(ref, () => list, [list])
  return (
    <FixedSizeList
      className={props.className}
      innerRef={props.innerRef}
      height={props.height}
      width={props.width}
      outerRef={outerRef}
      itemCount={list.count}
      itemSize={20}
      itemData={[...list.rows]}
      itemKey={(i) => i.toString()}
      innerElementType={InnerElement}
      overscanCount={8}
      onItemsRendered={(args) => {
        setRendered({
          startIndex: args.overscanStartIndex,
          stopIndex: args.overscanStopIndex,
        })
      }}
    >
      {Row}
    </FixedSizeList>
  )
})
