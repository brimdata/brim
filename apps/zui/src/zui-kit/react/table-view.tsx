import React, {forwardRef, useImperativeHandle, useMemo} from "react"
import {Provider} from "../../components/zed-table/context"
import {Grid} from "../../components/zed-table/grid"
import classNames from "classnames"
import {TableViewArgs} from "../core/table-view/types"
import {ReactAdapterProps} from "./types"
import {defaultTableViewState} from "../core/table-view/state"
import {TableViewApi} from "../core/table-view/table-view-api"
import {useStateControllers} from "./use-state-controllers"
import * as zed from "@brimdata/zed-js"

function useEnsureRecord(shape, values) {
  return useMemo(() => {
    if (shape instanceof zed.TypeRecord) {
      return [shape, values]
    } else {
      const wrappedValues = values.map((value) =>
        zed.createRecord({this: value})
      )
      const wrappedShape = wrappedValues[0]?.type
      return [wrappedShape, wrappedValues]
    }
  }, [shape, values])
}

export const TableView = forwardRef(function TableView(
  props: TableViewArgs & ReactAdapterProps,
  ref
) {
  const [shape, values] = useEnsureRecord(props.shape, props.values)
  const controllers = useStateControllers(props, defaultTableViewState)
  const args = {...props, ...controllers, shape, values}
  const api = useMemo(
    () => new TableViewApi(args),
    [
      values,
      shape,
      args.columnVisibleState.value,
      args.columnExpandedState.value,
    ]
  )

  api.update(args)

  useImperativeHandle(ref, () => api, [api])

  return (
    <Provider value={api}>
      <div
        role="grid"
        className={classNames("zed-table", {
          "zed-table--resizing": api.isResizing,
        })}
        ref={(node) => (api.element = node)}
      >
        <Grid />
      </div>
    </Provider>
  )
})
