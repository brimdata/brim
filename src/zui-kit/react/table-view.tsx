import React, {forwardRef, useImperativeHandle, useMemo} from "react"
import {Provider} from "../../components/zed-table/context"
import {Grid} from "../../components/zed-table/grid"
import classNames from "classnames"
import {TableViewArgs} from "../core/table-view/types"
import {ReactAdapterProps} from "./types"
import {defaultTableViewState} from "../core/table-view/state"
import {TableViewApi} from "../core/table-view/table-view-api"
import {useStateControllers} from "./use-state-controllers"

export const TableView = forwardRef(function TableView(
  props: TableViewArgs & ReactAdapterProps,
  ref
) {
  const controllers = useStateControllers(props, defaultTableViewState)
  const args = {...props, ...controllers}
  const api = useMemo(
    () => new TableViewApi(args),
    [
      args.values,
      args.shape,
      args.columnVisibleState.value,
      args.columnExpandedState.value,
    ]
  )

  api.update(args)

  useImperativeHandle(ref, () => api, [api])

  return (
    <Provider value={api}>
      <div
        className={classNames("zed-table", {
          "zed-table--resizing": api.isResizing,
        })}
        ref={(node) => (api.element = node)}
      >
        <Grid />
      </div>
    </Provider>
  )
})
