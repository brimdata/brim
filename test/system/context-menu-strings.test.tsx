import React from "react"
import {findAllByRole, screen} from "@testing-library/react"
import {SystemTest} from "./system-test"
import App from "src/js/components/App"

const system = new SystemTest("context-menu-strings")

describe("context menu tests", () => {
  beforeAll(async () => {
    system.render(<App />)
    await system.importFile("types.tsv")
    await screen.findAllByRole("cell")
  })

  describe("rightclick scalar strings", () => {
    test("-", scalarString("-"))
    test('"', scalarString('"'))
    test("'", scalarString("'"))
    test(",", scalarString(","))
    test(";", scalarString(";"))
    test("∫£œßü™", scalarString("∫£œßü™"))
    test("*", scalarString("*"))
    test('"*"', scalarString('"*"'))
    test("1.1.1.1", scalarString("1.1.1.1"))
  })
})

async function runTest(query: string, cellValue: string, menuText: string) {
  await system.runQuery(query)
  const cells = await screen.findAllByRole("cell")
  // This works with the empty string fields
  const cell = cells.find((c) => c.textContent === cellValue)
  system.rightClick(cell)
  system.click(await screen.findByText(menuText))
  expect(await findTableResults()).toMatchSnapshot()
}

function scalarString(value: string): () => Promise<void> {
  const query = `_path=="string" scalar!=null | cut id, scalar | sort id`
  return () => runTest(query, value, "Filter == value")
}

async function findTableResults() {
  const table = await screen.findByRole("table")
  const headers = await screen.findAllByRole("columnheader")
  const rows = await findAllByRole(table, "cell")
  return headers.concat(rows).map((r) => r.textContent)
}
