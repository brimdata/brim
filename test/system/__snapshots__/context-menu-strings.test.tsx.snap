// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`context menu tests rightclick scalar strings " 1`] = `
Array [
  "id",
  "scalar",
  "18",
  "\\"",
]
`;

exports[`context menu tests rightclick scalar strings "*" 1`] = `
Array [
  "id",
  "scalar",
  "27",
  "\\"*\\"",
]
`;

exports[`context menu tests rightclick scalar strings ' 1`] = `
Array [
  "id",
  "scalar",
  "19",
  "'",
]
`;

exports[`context menu tests rightclick scalar strings * 1`] = `
Array [
  "id",
  "scalar",
  "26",
  "*",
]
`;

exports[`context menu tests rightclick scalar strings , 1`] = `
Array [
  "id",
  "scalar",
  "20",
  ",",
  "21",
  ",",
]
`;

exports[`context menu tests rightclick scalar strings ; 1`] = `
Array [
  "id",
  "scalar",
  "22",
  ";",
]
`;

exports[`context menu tests rightclick scalar strings - 1`] = `
Array [
  "id",
  "scalar",
  "16",
  "-",
]
`;

exports[`context menu tests rightclick scalar strings 1.1.1.1 1`] = `
Array [
  "id",
  "scalar",
  "28",
  "1.1.1.1",
]
`;

exports[`context menu tests rightclick scalar strings ∫£œßü™ 1`] = `
Array [
  "id",
  "scalar",
  "23",
  "∫£œßü™",
]
`;
